api/src/database/run-ast/utils/merge-with-parent-items.ts

```diff
@@ -16,39 +16,81 @@ export function mergeWithParentItems(
 	const parentItems = clone(toArray(parentItem));
 
 	if (nestedNode.type === 'm2o') {
-		for (const parentItem of parentItems) {
-			const itemChild = nestedItems.find((nestedItem) => {
-				return (
-					nestedItem[schema.collections[nestedNode.relation.related_collection!]!.primary] ==
-					parentItem[nestedNode.relation.field]
-				);
-			});
+		const parentsByForeignKey = new Map();
```

Comment:
I think it wouldn't hurt to be explicit regarding the datatypes of the maps, to get type safety and directly see what map will be build. I assume it would be this:
```ts
import type { PrimaryKey } from '@directus/types';

const parentsByForeignKey: Map<PrimaryKey, Item[]> = new Map();
```
(the name of the type is a bit confusing here, but I guess that's how it's done so far).

similarly for the other new map `parentsByPrimaryKey`. 

```ts
const parentsByPrimaryKey: Map<PrimaryKey, Item> = new Map();
```